pkg/util/env/env.go

```diff
@@ -0,0 +1,134 @@
+	"k8s.io/kubernetes/pkg/fieldpath"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+type ResourceStore struct {
+	SecretStore    map[string]*kapi.Secret
+	ConfigMapStore map[string]*kapi.ConfigMap
+}
+
+func NewResourceStore() *ResourceStore {
+	return &ResourceStore{
+		SecretStore:    make(map[string]*kapi.Secret),
+		ConfigMapStore: make(map[string]*kapi.ConfigMap),
+	}
+}
+
+func GetEnvVarSecretRefValue(client kclientset.Interface, namespace string, store *ResourceStore, secretSelector *kapi.SecretKeySelector) (string, error) {
+	secret, ok := store.SecretStore[secretSelector.Name]
+	if !ok {
+		var err error
+		secret, err = client.Core().Secrets(namespace).Get(secretSelector.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		store.SecretStore[secretSelector.Name] = secret
+	}
+	if data, ok := secret.Data[secretSelector.Key]; ok {
+		return string(data), nil
+	}
+	return "", fmt.Errorf("key %s not found in secret %s", secretSelector.Key, secretSelector.Name)
+
+}
+
+func getSecretRefValue(f *clientcmd.Factory, store *ResourceStore, secretSelector *kapi.SecretKeySelector) (string, error) {
+	kubeClient, err := f.ClientSet()
+	if err != nil {
+		return "", err
+	}
+	namespace, _, err := f.DefaultNamespace()
+	if err != nil {
+		return "", err
+	}
+	return GetEnvVarSecretRefValue(kubeClient, namespace, store, secretSelector)
+}
+
+func GetEnvVarConfigMapRefValue(client kclientset.Interface, namespace string, store *ResourceStore, configMapSelector *kapi.ConfigMapKeySelector) (string, error) {
+	configMap, ok := store.ConfigMapStore[configMapSelector.Name]
+	if !ok {
+		var err error
+		configMap, err = client.Core().ConfigMaps(namespace).Get(configMapSelector.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		store.ConfigMapStore[configMapSelector.Name] = configMap
+	}
+	if data, ok := configMap.Data[configMapSelector.Key]; ok {
+		return string(data), nil
+	}
+	return "", fmt.Errorf("key %s not found in config map %s", configMapSelector.Key, configMapSelector.Name)
+}
+
+func getConfigMapRefValue(f *clientcmd.Factory, store *ResourceStore, configMapSelector *kapi.ConfigMapKeySelector) (string, error) {
+	kubeClient, err := f.ClientSet()
+	if err != nil {
+		return "", err
+	}
+	namespace, _, err := f.DefaultNamespace()
+	if err != nil {
+		return "", err
+	}
+	return GetEnvVarConfigMapRefValue(kubeClient, namespace, store, configMapSelector)
+}
+
+func GetEnvVarFieldRef(obj runtime.Object, from *kapi.EnvVarSource) (string, error) {
+	return fieldpath.ExtractFieldPathAsString(obj, from.FieldRef.FieldPath)
+}
+
+func GetEnvVarResourceFieldRef(from *kapi.EnvVarSource, c *kapi.Container) (string, error) {
+	return fieldpath.InternalExtractContainerResourceValue(from.ResourceFieldRef, c)
+}
+
+func GetEnvVarRefValue(f *clientcmd.Factory, store *ResourceStore, from *kapi.EnvVarSource, obj runtime.Object, c *kapi.Container) (string, error) {
```

Comment:
why isn't there an equivalent of this for the build logic to call, instead of including all the switching in the build controller?